### PR TITLE
Duplicate timestamp and input-sorting changes

### PR DESCRIPTION
**Pull request by andyleiserson**
IPAv3 sorts the input records, so add sorting of input records to ipa_in_the_clear, and remove it from input drivers.

IPAv3 also shuffles the input records, so if there are duplicate timestamps among a single user's records, the output may be non-deterministic. To avoid end-to-end test failures, modify the test data generator to avoid generating duplicate timestamps. Add a directed test to exercise duplicate timestamps.

Possibly a fix for #967.

### CODE REVIEW

**Comment by akoshelev** [comment]
@eriktaubeneck this may solve #967

File: ipa-core/src/test_fixture/ipa.rs [hunk 1]
```diff
@@ -41,13 +43,31 @@ impl GroupingKey for TestRawDataRecord {
     }
 }
 
+/// Insert `record` into `user_records`, maintaining timestamp order.
+///
+/// If there are existing records with the same timestamp, inserts the new record
+/// randomly in any position that maintains timestamp order.
+fn insert_sorted(user_records: &mut Vec<TestRawDataRecord>, record: TestRawDataRecord) {
```
Comment (akoshelev) [inline ipa-core/src/test_fixture/ipa.rs:50]: have you had a chance to test how long it takes to generate ~20M events with this change? I don't expect it to be terrible, but it would be great to know that for sure

File: ipa-core/src/test_fixture/event_gen.rs [hunk 2]
```diff
@@ -166,15 +170,21 @@ impl<R: Rng> EventGenerator<R> {
...
-        // Generate a new random timestamp between [0..`max_timestamp`).
-        // This means the generated events must be sorted by timestamp before being
-        // fed into the IPA protocols.
-        let current_ts = self.rng.gen_range(0..self.config.max_timestamp.get());
+    fn gen_event(&mut self, idx: usize) -> TestRawDataRecord {
+        let user_id = self.users[idx].user_id;
+
+        // Generate a new random timestamp between [0..`max_timestamp`) and distinct from
+        // already-used timestamps.
+        let current_ts = loop {
```
Comment (akoshelev) [inline ipa-core/src/test_fixture/event_gen.rs:182]: this has some potential to stall, so maybe we could check if `used_timestamp` size == max_timestamp?

**Review by akoshelev — APPROVED** [review]

File: ipa-core/src/test_fixture/ipa.rs (hunk 1 above)
Comment (andyleiserson) [reply ipa-core/src/test_fixture/ipa.rs:50]: Running the revised event generation and `ipa_in_the_clear` with max_events_per_user up to 1,000 takes 5-10% longer than the previous version, which seems tolerable to me. With max_events_per_user of 10,000 the increase is more like 40%.
